test/tools/uri_spec_runner.ts

```diff
@@ -0,0 +1,304 @@
+    source: string;
+    mechanism: string;
+    mechanism_properties: Record<string, any>;
+  };
+}
+
+export function executeUriValidationTest(
+  test: UriTest | AuthTest,
+  shouldNotThrowOnWarn = false
+): void {
+  const knownTestKeys = [
+    'description',
+    'uri',
+    'valid',
+    'warning',
+    'hosts',
+    'auth',
+    'options',
+    'credential'
+  ];
+  expect(knownTestKeys).to.include.members(Object.keys(test));
+
+  const errorMessage = `"${test.uri}"`;
+
+  const valid = test.valid && (!(test as UriTest).warning || shouldNotThrowOnWarn);
+
+  if (!valid) {
+    try {
+      new MongoClient(test.uri);
+      expect.fail(`Expected "${test.uri}" to be invalid${test.valid ? ' because of warning' : ''}`);
+    } catch (err) {
+      if (err instanceof TypeError) {
+        expect(err).to.have.property('code').equal('ERR_INVALID_URL');
+      } else if (
+        // most of our validation is MongoParseError, which does not extend from MongoAPIError
+        !(err instanceof MongoParseError) &&
+        // the rest of our validation extends from MongoAPIError
+        !(err instanceof MongoAPIError) &&
+        // mongodb-connection-string-url does not export its MongoParseError so we can't check for it directly
+        err.name !== 'MongoParseError'
+      ) {
+        throw err;
+      }
+    }
+    return;
+  }
+
+  const client = new MongoClient(test.uri);
+  const options = client.options;
+  expect(options, errorMessage).to.be.an('object').that.is.not.empty;
+
+  if ('hosts' in test && test.hosts != null) {
+    for (const [index, { host, port }] of test.hosts.entries()) {
+      const actualHost = options.hosts[index];
+      if (actualHost.host == null && actualHost.socketPath == null) {
+        expect.fail('Expected host to define "host" or "socketPath" properties');
+      }
+      if (actualHost.host != null) {
+        expect(actualHost, errorMessage).property('host').to.equal(host);
+      } else {
+        expect(actualHost, errorMessage).property('socketPath').to.equal(host);
+      }
+
+      if (port != null) expect(actualHost).property('port').to.equal(port);
+    }
+  }
+
+  // depending on whether this is a UriTest or an AuthTest, credential option values are
+  // defined in test.credential or test.auth, respectively, and additional properties
+  // to test are defined in test.options or mixed into test.credential, respectively
+  let credentialsToTest: { source?: string; password?: string; username?: string } = {};
+  let optionsToTest: Record<string, any> = 'options' in test ? test.options || {} : {};
+
+  if (!('options' in test) && test.credential != null) {
+    // handle AuthTest credential testing
+    const credentialOptions = [
+      'username',
+      'password',
+      'source',
+      'mechanism',
+      'mechanism_properties'
+    ];
+    expect(test).property('credential').to.have.all.keys(credentialOptions);
+
+    optionsToTest = (({ mechanism, mechanism_properties }) => ({
```

Comment:
this is clever but it took me a bit to figure out what's happening.  Can we just assign the values normally? 

```typescript
optionsToTest = {
  mechanism: test.credential.mechanism,
  mechanism_properties: test.credential.mechanism_properties
}
```

Reply:
sure, we can be boring with them if the team wants 😄  it just seemed like the one rare opportunity to use an immediately invoked function to keep things concise @nbbeeken - any strong feelings here?

Reply:
Is the idea here to pull apart the test.credentials object for readability of an `optionsToTest` variable or is it important that we're dropping the other options that aren't mechanism/mechanism_properties?

Reply:
the options to test have to be *just* the mechanism and mechanism_properties, the other options have to be handled in the credentials block below; not sure I'm answering your question, but if not, could you clarify what you mean?

Reply:
Discussed in slack the alternatives: what we have here works for us